src/client/Client.cc

```diff
@@ -8438,6 +8454,29 @@ int Client::_open(Inode *in, int flags, mode_t mode, Fh **fhp,
     req->head.args.open.old_size = in->size;   // for O_TRUNC
     req->set_inode(in);
     result = make_request(req, perms);
+
+    /*
+     * Clients that have delegations enabled will wait on minimal caps, just
+     * to ensure that other clients holding delegations return theirs first.
+     */
+    if (deleg_timeout && result == 0) {
+      int need = 0, have;
+
+      if (cmode & CEPH_FILE_MODE_WR)
+        need |= CEPH_CAP_FILE_WR;
+      if (cmode & CEPH_FILE_MODE_RD)
+        need |= CEPH_CAP_FILE_RD;
+
+      result = get_caps(in, need, want, &have, -1);
+      if (result < 0) {
+	ldout(cct, 1) << "Unable to get caps after open of inode " << *in <<
+			  " . Denying open: " <<
+			  cpp_strerror(result) << dendl;
+	in->put_open_ref(cmode);
+      } else {
+	put_cap_ref(in, need);
```

Comment:
put_cap_ref(in, have);

Reply:
I'm not sure that's right. Sadly, get_caps doesn't have a comment header to explain what the semantics should be, but we do have this:

`    if (!waitfor_caps && !waitfor_commit) {                                     
      if ((have & need) == need) {                                              
        int revoking = implemented & ~have;                                     
        ldout(cct, 10) << "get_caps " << *in << " have " << ccap_string(have)   
                 << " need " << ccap_string(need) << " want " << ccap_string(want)
                 << " revoking " << ccap_string(revoking)                       
                 << dendl;                                                      
        if ((revoking & want) == 0) {                                           
          *phave = need | (have & want);                                        
          in->get_cap_ref(need);                                                
          return 0;                                                             
        }                                                                       
      }                                                                         
`

It looks like we only tape cap references on "need", not what ends up in "have".

Reply:
Oh, you are right